### PR TITLE
Implementation of extended translation functionality, as well as major refactoring and removal of old, unused code and tests

### PR DESCRIPTION
**Pull request by KuhnMn**
Translation from and to Cirq, Q#, Amazon Braket, and Quirk was implemented using third-party frameworks pytket, Pennylane, and staq.
Translation from Q# and Braket IR was implemented manually using mappings.

### CODE REVIEW

**Review by mbeisel — CHANGES_REQUESTED** [review]
needs some refactoring and the requirements need to be checked again

File: test/converter_tests/braket_test.py [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+        org_circuit.sample(Observable.Y(), 1)
+        org_circuit.sample(Observable.Z(), 2)
+        org_counts = simulate_braket(org_circuit)
+        wrapper.import_braket_ir(org_circuit.to_ir().json(indent=4))
+        t_circuit = wrapper.export_qiskit()
+        t_counts = simulate_qiskit(t_circuit)
+        similarity = intersection(org_counts, t_counts)
+        self.assertGreater(similarity, 0.90)
+
+
```
Comment (FelixTruger) [inline test/converter_tests/braket_test.py:107]: Why are there so many newlines? (maybe run black to apply some reasonable code style)

File: frontend/src/app/analyze/analyze.component.html [hunk 2]
```diff
@@ -49,22 +49,39 @@
...
+                    </mat-option>
+                    <mat-option
+                        matTooltip="Pyquil program can be exported in the backend. The export of Pyquil skeleton code is experimential."
+                        matTooltipClass="tooltip-large" value="Pyquil">
+                        Pyquil
+                    </mat-option>
+                    <mat-option matTooltip="No rzz gate; Custom Gate definitions will be added for non qsharp native gates."
+                        matTooltipClass="tooltip-large"  value="Qsharp">
+                        QSharp
+                    </mat-option>
```
Comment (FelixTruger) [inline frontend/src/app/analyze/analyze.component.html:84]: Indentation seems off

File: conversion/converter/command_converter.py [hunk 3]
```diff
@@ -9,16 +9,34 @@
...
 from pyquil.gates import NOP, MEASURE
+import cirq
+from cirq import Circuit
+from braket.circuits.circuit import Circuit as BRCircuit
 
-def pyquil_commands_to_program(commands: str) -> QuantumCircuit:    
+
+def pyquil_commands_to_program(commands: str) -> Program:
     exec(commands)
     val = eval("p")
```
Comment (FelixTruger) [inline conversion/converter/command_converter.py:19]: From Martin's comments: "the use of eval and exec is not safe (is there an alternative solution?)"

**Review by FelixTruger — COMMENTED** [review]
Some more comments and hints for improvements

File: test/id_comparison.py [hunk 4]
```diff
@@ -56,7 +56,8 @@ def simulate_circuit(circuit: QuantumCircuit):
     figure.savefig("./test/results/plot.pdf")
 
 if __name__ == "__main__":
-    # circuit = bernstein_vazirani_general_qiskit_integer(4, 8, True)
```
[the remark this answers is not in the document]
Comment (salmma) [reply test/id_comparison.py:59]: yes or no?

**Comment by salmma** [comment]
please try to merge the open PRs in your branch and check if everything is still running as expected

**Review by salmma — CHANGES_REQUESTED** [review]
see https://github.com/UST-QuAntiL/QuantumTranspiler/pull/19#issuecomment-1351630432
+
Review about conversion testing using the conversion feature of the UI:
In the following, cases are described where I got errors.

**Case 1:**
1. https://raw.githubusercontent.com/UST-QuAntiL/cirq-service/main/Sample%20Implementations/ciruit_json.json
2. to QSharp
3. to any other language besides OpenQASM throws errors

**Case 2:**
1. same circuit as in Case 1, step 1
2. to Quirk
3. to any other language throws errors

**Case 3:**
1. https://raw.githubusercontent.com/UST-QuAntiL/braket-service/main/Sample%20Implementations/circuit_braket.py 
2. to any other language throws errors

**Case 4:**

[Sample circuit](https://raw.githubusercontent.com/UST-QuAntiL/qsharp-service/main/Sample%20Implementations/circuit_qsharp_params.qs) in qsharp-service cannot be translated into other languages, because of the parameters, I assume?

File: api/api_service.py [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+        return f"General error while converting to {language_output}: {str(e)}", 500
+    return output
+
+
+@app.route("/unroll", methods=["Post"])
+def unroll():
+    data = request.json
+    architecture = data["option"]
+    circuit = data["circuit"]
+    is_custom_export = data["isExpert"]
```
[the remark this answers is not in the document]
Comment (KuhnMn) [reply api/api_service.py:129]: So this is a bit confusing but hard to change immediately. Inside the request, this parameter is named "isExpert". However, Martin and Felix told me that this name is rather confusing, so I wanted to change it to something that explains what it is better, which is a custom export, where the user can choose the export format themselves. However, changing that in the request format would mean also changing it in the NISQ analyzer and all other services which use the QuantumTranspiler, which Martin and Felix told me I should do at a later point. This is why, in the code, it is named custom_export, which explains it better and in the JSON it's named "isExpert" for compatibility.

**Comment by KuhnMn** [comment]
Regarding the cases of the last comment:

1. This is due to Cirq and pytket not being able to import and braket not implementing the "Reset" operation. When using pytket to translate from OpenQASM to QSharp, it adds reset operations after every measurement for some reason. So, when trying to translate back, it works for QASM, since I did this one manually, but everything else which is done by pytket or cirq doesn't work. I adjusted the code for Quil, so that it now also supports the reset operation, but for everything else I'd need to implement the converter from scratch to fix this.
2. Was a bug with the Quirk converter, which is now fixed.
3. So this one is because the function is never executed. How the conversion works is that the Transpiler executes the code given and then looks for a variable named "c" and evaluates that (This is kind of hidden, but when hovering over the language in the drop down menu you can see which name the circuit has to have). However, since the method is never executed and the result is never written into a variable "c", no circuit is found. If you add for example the line `c = get_circuit(1)` you are able to translate it.
4. Yes, this is because of the parameter. If you remove the parameter and change it to a value like 1.0, the transpiler will be able to translate it.

Lastly, I assume "merge the open PRs in your branch" means merging all the dependabot PRs into my branch? Or did I misunderstand this?

**Comment by salmma** [comment]
> Lastly, I assume "merge the open PRs in your branch" means merging all the dependabot PRs into my branch? Or did I misunderstand this?
exactly

**Comment by KuhnMn** [comment]
This should be all relevant PRs merged. the remaining PRs are either for python dependencies that I removed due to the dependency cleanup I did a while back or the js dependency karma, which I had to do manually because the PR broke everything for some reason. I also did an npm audit fix in the end to solve some of the dependency issues. The only thing I had to revert was the PR for the angular-core dependency since that one broke the entire dependency tree. I tried to also update this dependency manually and with npm, but whatever I tried the whole thing collapsed, so I decided to keep it as it was.

**Review by github-fake2 — APPROVED** [review]

**Review by salmma — APPROVED** [review]

**Comment by salmma** [comment]
simply ignore github-fake2 user 😂

**Comment by salmma** [comment]
@KuhnMn could you please check and merge the open dependency PRs? :)

**Comment by KuhnMn** [comment]
> @KuhnMn could you please check and merge the open dependency PRs? :)

I merged all PRs that were compatible with the translation branch. All other PRs either refer to dependencies that are no longer part of the project or are js dependencies that have been changed by the past updates I have done. I tried to make dependabot recreate these for the new branch, but that didn't work. Should I close them manually, or wait for the merge to see if dependabot can work with them afterward?

**Review by mbeisel — APPROVED** [review]

**Pull request merged**